Implement a server-side cursor to improve performance when fetching large volumes of data. #5797

khushboovashi opened the pull request.

akshay-joshi reviewed (changes requested):
**GUI-related comments:**
1) Select 'Use server cursor' from the query tool. Run "select 1" query. Clicking on the Commit button gets the error message " the cursor is closed" and on Rollback "DECLARE "CURSOR:CONN:961921" CURSOR FOR ROLLBACK;" Disable those buttons in server cursor mode.
2) Set "Execute with server side cursor?" To true. Select any table and click on View/Edit data. It executes the query, but again commit and rollback buttons get enabled. Now try following:

- Try to close the tab. It will not.
- Pagination shows Page no 1 of 0.
- Click on Next page or Previous page, nothing happens, and the data is lost.    

**Code-related comments:**
1) Add this new 'Use server cursor?" preference to the preferences.rst. The screenshot will be updated after Aditya's PR.
2) Fix 3 small code smells in sqleditor/__init__.py

khushboovashi commented on docs/en_US/query_tool.rst:
```diff
@@ -558,3 +558,36 @@ To execute a macro, simply select the appropriate shortcut keys, or select it fr
+* Alternatively, you can enable it on a per-session basis via the Query Tool’s Execute menu.
+
+.. image:: images/query_tool_server_cursor_execute_menu.png
+   :alt: Query Tool Manage Macros Clear row confirmation
```
Done

khushboovashi commented on web/pgadmin/utils/driver/psycopg3/cursor.py:
```diff
@@ -17,7 +17,8 @@
-from psycopg import Cursor as _cursor, AsyncCursor as _async_cursor
+from psycopg import (Cursor as _cursor, AsyncCursor as _async_cursor,
+                     AsyncServerCursor as _async_server_cursor)
 from typing import Any, Sequence
```
This is old import but I have removed.

khushboovashi commented on web/pgadmin/utils/driver/psycopg3/connection.py:
```diff
@@ -732,10 +742,14 @@ def __cursor(self, server_cursor=False, scrollable=False):
                 # Providing name to cursor will create server side cursor.
                 cursor_name = "CURSOR:{0}".format(self.conn_id)
+                self.conn.server_cursor_factory = AsyncDictServerCursor
+                # self.__reset_auto_commit(False)
```
Done

khushboovashi commented on web/pgadmin/tools/sqleditor/utils/query_tool_preferences.py:
```diff
@@ -80,6 +80,17 @@ def register_query_tool_preferences(self):
+                         ' server-side cursor after query execution.'
+                         ' This will be helpful to transfer to the client only'
+                         ' a controlled amount of data, so a large dataset can'
+                         ' be examined without being kept entirely in memory.')
```
Done, though I have taken the older text from psycopg doc.

khushboovashi commented on web/pgadmin/tools/sqleditor/tests/test_server_cursor.py:
```diff
@@ -0,0 +1,118 @@
+# -*- coding: utf-8 -*-
```
done

khushboovashi commented on web/pgadmin/tools/sqleditor/utils/query_tool_preferences.py:
```diff
@@ -80,6 +80,17 @@ def register_query_tool_preferences(self):
 
+    self.server_cursor = self.preference.register(
+        'Options', 'server_cursor',
+        gettext("Execute with server side cursor?"), 'boolean', False,
```
done

khushboovashi commented on web/pgadmin/tools/sqleditor/static/js/components/sections/MainToolBar.jsx:
```diff
@@ -634,6 +638,8 @@ export function MainToolBar({containerRef, onFilterClick, onManageMacros, onAddT
         <PgMenuItem hasCheck value="auto_rollback" checked={checkedMenuItems['auto_rollback']}
           onClick={checkMenuClick}>{gettext('Auto rollback on error?')}</PgMenuItem>
+        <PgMenuItem hasCheck value="server_cursor" checked={checkedMenuItems['server_cursor']}
+          onClick={checkMenuClick}>{gettext('Use server cursor')}</PgMenuItem>
```
done

khushboovashi commented:
> **GUI-related comments:**
> 
> 1. Select 'Use server cursor' from the query tool. Run "select 1" query. Clicking on the Commit button gets the error message " the cursor is closed" and on Rollback "DECLARE "CURSOR:CONN:961921" CURSOR FOR ROLLBACK;" Disable those buttons in server cursor mode.

Fixed the issue. I have disabled the buttons right now in View/Edit data but that is debatable, we will discuss that in the review.

> 2. Set "Execute with server side cursor?" To true. Select any table and click on View/Edit data. It executes the query, but again commit and rollback buttons get enabled. Now try following:
> 
> * Try to close the tab. It will not.
Fixed
> * Pagination shows Page no 1 of 0.
Fixed
> * Click on Next page or Previous page, nothing happens, and the data is lost.
Didn't reproduce
> 
> **Code-related comments:**
> 
> 1. Add this new 'Use server cursor?" preference to the preferences.rst. The screenshot will be updated after Aditya's PR.
Fixed
> 2. Fix 3 small code smells in sqleditor/**init**.py

akshay-joshi reviewed (changes requested):
Update the preferences screenshot.
